Cache-Hash-Data-Structure.adoc

```diff
@@ -0,0 +1,209 @@
+= Hash Data Structure
+
+Provide a Hash data structure to Infinispan users. The Hash is a flexible identifiable grouping of keys mapping to values.
+In this context, _flexible_ means it can have arbitrary mappings without needing a schema beforehand; _identifiable group_
+means the Hash structure receives a unique name and retrieves the complete key-value mapping.
+
+The Hash provides a general approach for users to represent their data structures using Infinispan. For example, by
+employing this, users can avoid using key prefixes to create an object in Cache.
+
+== Public API
+
+The work is for the new API only. The Hash data structure has the format of:
+
+[source,java]
+.Hash.java
+----
+public interface Hash<K, V> { // <1>
+
+   String getName();
+
+   V get(K key);
+
+   V put(K key, V value);
+
+   V remove(K key);
+
+   boolean containsKey(K key);
+
+   Set<K> keySet();
+
+   Collection<V> values();
+
+   Set<Map.Entry<K, V>> entrySet();
+
+   void clear();
+
+   int size();
+}
+----
+
+<1> The Hash data structure is a generic interface over `K` and `V`. We do not extend `Map<K, V>`, instead we add methods
+as needed.
+
+A `Hash<?, ?>` structure is local to a node. Meaning that operations applied to the local copy must be submitted
+through the public API before taking effect. After that, the `Hash<?, ?>` is distributed/replicated according to
+the configuration.
+
+We follow a state-based CRDT approach, merging the local Hash with the server's Hash. Since the `Hash<?, ?>` structures
```

Comment:
how are we going to merge when network partitions happen? Are we going to support cross-site?